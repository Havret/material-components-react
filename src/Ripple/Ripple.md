### Bounded

```js
<div className="demo-surface">
    <Ripple        
        render={({innerRef, ...rest}) =>
            (<div {...rest} ref={innerRef}>Interact with me</div>)}
    />
</div>
```

### Unbounded

```js
<div className="demo-surface">
    <Ripple
        unbounded={true}               
        render={({innerRef, ...rest}) =>
            (<div {...rest} ref={innerRef}>Interact with me</div>)}
    />
</div>
```

### Theme Styles

```js
<div className="demo-surface">
    <Ripple
        primary={true}               
        render={({innerRef, ...rest}) =>
            (<div {...rest} ref={innerRef}>Primary</div>)}
    />
    
    <Ripple
            secondary={true}               
            render={({innerRef, ...rest}) =>
                (<div {...rest} ref={innerRef}>Secondary</div>)}
        />
</div>
```