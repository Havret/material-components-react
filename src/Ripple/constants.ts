const cssClasses = {
    // Ripple is a special case where the "root" component is really a "mixin" of sorts,
    // given that it's an 'upgrade' to an existing component. That being said it is the root
    // CSS class that all other CSS classes derive from.
    ROOT: 'mdc-ripple-upgraded',
    UNBOUNDED: 'mdc-ripple-upgraded--unbounded',
    BG_FOCUSED: 'mdc-ripple-upgraded--background-focused',
    BG_ACTIVE_FILL: 'mdc-ripple-upgraded--background-active-fill',
    FG_ACTIVATION: 'mdc-ripple-upgraded--foreground-activation',
    FG_DEACTIVATION: 'mdc-ripple-upgraded--foreground-deactivation',
};

const strings = {
    VAR_FG_SIZE: '--mdc-ripple-fg-size',
    VAR_LEFT: '--mdc-ripple-left',
    VAR_TOP: '--mdc-ripple-top',
    VAR_FG_SCALE: '--mdc-ripple-fg-scale',
    VAR_FG_TRANSLATE_START: '--mdc-ripple-fg-translate-start',
    VAR_FG_TRANSLATE_END: '--mdc-ripple-fg-translate-end',
};

const numbers = {
    PADDING: 10,
    INITIAL_ORIGIN_SCALE: 0.6,
    DEACTIVATION_TIMEOUT_MS: 225, // Corresponds to $mdc-ripple-translate-duration (i.e. activation animation duration)
    FG_DEACTIVATION_MS: 150, // Corresponds to $mdc-ripple-fade-out-duration (i.e. deactivation animation duration)
};

export { cssClasses, strings, numbers };