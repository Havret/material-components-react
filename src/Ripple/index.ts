export { Ripple, RippleProps, RippleComponentProps, defaultBoundingRect } from './Ripple';