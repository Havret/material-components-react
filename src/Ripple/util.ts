import { MouseEvent, TouchEvent } from 'react';

let _supportsCssVariables: boolean;

function supportsCssVariables(forceRefresh: boolean = false) {
    if (typeof _supportsCssVariables === 'boolean' && !forceRefresh) {
        return _supportsCssVariables;
    }

    const supportsFunctionPresent = CSS && typeof CSS.supports === 'function';
    if (!supportsFunctionPresent) {
        return;
    }

    const explicitlySupportsCssVars = CSS.supports('--css-vars', 'yes');
    // See: https://bugs.webkit.org/show_bug.cgi?id=154669
    // See: README section on Safari
    const weAreFeatureDetectingSafari10plus = (
        CSS.supports('(--css-vars: yes)') &&
        CSS.supports('color', '#00000000')
    );

    if (explicitlySupportsCssVars || weAreFeatureDetectingSafari10plus) {
        _supportsCssVariables = !detectEdgePseudoVarBug();
    } else {
        _supportsCssVariables = false;
    }
    return _supportsCssVariables;
}

function detectEdgePseudoVarBug() {
    // Detect versions of Edge with buggy var() support
    // See: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/11495448/
    const document = window.document;
    const node = document.createElement('div');
    node.className = 'mdc-ripple-surface--test-edge-var-bug';
    document.body.appendChild(node);

    // The bug exists if ::before style ends up propagating to the parent element.
    // Additionally, getComputedStyle returns null in iframes with display: "none" in Firefox,
    // but Firefox is known to support CSS custom properties correctly.
    // See: https://bugzilla.mozilla.org/show_bug.cgi?id=548397
    const computedStyle = window.getComputedStyle(node);
    const hasPseudoVarBug = computedStyle !== null && computedStyle.borderTopStyle === 'solid';
    node.remove();
    return hasPseudoVarBug;
}

let _supportsPassive: boolean;

function applyPassive(globalObj: Window = window, forceRefresh: boolean = false): boolean | AddEventListenerOptions {
    if (_supportsPassive === undefined || forceRefresh) {
        let isSupported = false;
        const options = Object.defineProperty({}, 'passive', {
            get() { isSupported = true; },
        });

        // tslint:disable-next-line:no-empty
        const noop = () => {};
        globalObj.document.addEventListener('testPassiveEventSupport', noop, options);
        globalObj.document.removeEventListener('testPassiveEventSupport', noop, options);

        _supportsPassive = isSupported;
    }

    return _supportsPassive ? {passive: true} : false;
}

type Event = TouchEvent<{}> | MouseEvent<{}>;
function getNormalizedEventCoords(event: Event, pageOffset: { x: number, y: number }, clientRect: ClientRect) {
    const {x, y} = pageOffset;
    const documentX = x + clientRect.left;
    const documentY = y + clientRect.top;

    let normalizedX;
    let normalizedY;

    // Determine touch point relative to the ripple container.
    if (isTouchEvent(event)) {
        normalizedX = event.changedTouches[0].pageX - documentX;
        normalizedY = event.changedTouches[0].pageY - documentY;
    } else {
        normalizedX = event.pageX - documentX;
        normalizedY = event.pageY - documentY;
    }

    return {x: normalizedX, y: normalizedY};
}

function isTouchEvent(event: TouchEvent<{}> | MouseEvent<{}>): event is TouchEvent<{}> {
    return event.type === 'touchstart';
}

export { supportsCssVariables, applyPassive, getNormalizedEventCoords };