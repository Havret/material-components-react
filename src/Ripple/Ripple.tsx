import * as cx from 'classnames';
import * as React from 'react';
import { CSSProperties, EventHandler, MouseEvent, TouchEvent, FocusEvent, ReactNode } from 'react';

import { cssClasses, numbers, strings } from './constants';
import { applyPassive, getNormalizedEventCoords } from './util';

const defaultBoundingRect: ClientRect = {
    height: 0,
    width: 0,
    top: 0,
    left: 0,
    right: 0,
    bottom: 0
};

enum RippleMode {
    none, activated, deactivated
}

interface RippleState {
    eventCoords: {
        x: number,
        y: number,
    };
    frame: {
        width: number;
        height: number;
    };
    mode: RippleMode;
    focused: boolean;
}

interface RippleComponentProps {
    onTouchStart: EventHandler<TouchEvent<{}>>;
    onTouchEnd: EventHandler<TouchEvent<{}>>;
    onMouseDown: EventHandler<MouseEvent<{}>>;
    onMouseUp: EventHandler<MouseEvent<{}>>;
    onBlur: EventHandler<FocusEvent<{}>>;
    onFocus: EventHandler<FocusEvent<{}>>;
    ref?: (ref: Element | null) => void;
    className: string;
    style: CSSProperties;
}

interface RippleProps {
    /**
     * If `true` adds base styles for a ripple surface.
     */
    surface?: boolean;
    /**
     * If `true` styles ripple surface using theme's primary colors.
     */
    primary?: boolean;
    /**
     * If `true` styles ripple surface using theme's secondary colors.
     */
    secondary?: boolean;
    /**
     * If `true` lets a ripple grow outside of its bounds, just like on checkboxes or radio buttons.
     */
    unbounded?: boolean;
    /**
     * If `true`, the ripple will be disabled.
     */
    disabled?: boolean;
    /**
     * Returns the ClientRect for the surface.
     * @returns {ClientRect}
     */
    computeBoundingRect?: () => ClientRect;
    render: ((props: RippleComponentProps) => React.ReactNode);
}

class Ripple extends React.Component<RippleProps, RippleState> {
    static defaultProps = {
        surface: true,
        primary: false,
        secondary: false,
        unbounded: false,
        disabled: false,
    };

    state = {
        eventCoords: {x: 0, y: 0},
        frame: {width: 0, height: 0},
        mode: RippleMode.none,
        focused: false
    };

    private _innerRef: Element | null;
    private _layoutFrame: number;
    private _activationAnimationHasEnded: boolean;
    private _activationHasEnded: boolean;

    // https://github.com/Microsoft/TypeScript/issues/842
    // tslint:disable:no-any
    private _activationTimer: any;
    private _fgDeactivationRemovalTimer: any; // tslint:enable:no-any

    componentDidMount(): void {
        window.addEventListener('resize', this.handleResize);

        const {width, height} = this.getBoundingClientRect();
        this.setState({frame: {width, height}});
    }

    componentWillUnmount(): void {
        window.removeEventListener('resize', this.handleResize);
    }

    componentDidUpdate(prevProps: Readonly<RippleProps>, prevState: Readonly<RippleState>) {
        const {mode} = this.state;

        if (mode === RippleMode.activated) {
            clearTimeout(this._activationTimer);
            clearTimeout(this._fgDeactivationRemovalTimer);

            this._activationAnimationHasEnded = false;

            this._activationTimer = setTimeout(() => {
                this._activationAnimationHasEnded = true;
                this.deactivate();
            }, numbers.DEACTIVATION_TIMEOUT_MS);
        } else if (mode === RippleMode.deactivated) {
            clearTimeout(this._fgDeactivationRemovalTimer);
            this._fgDeactivationRemovalTimer = setTimeout(() => {
                this.setState({mode: RippleMode.none});
            }, numbers.FG_DEACTIVATION_MS);
        }
    }

    render() {
        const {unbounded, surface, primary, secondary, computeBoundingRect} = this.props;
        const {frame, mode, focused} = this.state;
        const {height, width} = frame;

        const classNames = cx(cssClasses.ROOT, {
            [cssClasses.UNBOUNDED]: unbounded,
            [cssClasses.FG_ACTIVATION]: mode === RippleMode.activated,
            [cssClasses.FG_DEACTIVATION]: mode === RippleMode.deactivated,
            [cssClasses.BG_FOCUSED]: focused,
            [cssClasses.SURFACE]: surface,
            [cssClasses.SURFACE_PRIMARY]: primary,
            [cssClasses.SURFACE_SECONDARY]: secondary
        });

        const maxDim = Math.max(height, width);
        const surfaceDiameter = Math.sqrt(Math.pow(width, 2) + Math.pow(height, 2));
        const initialSize = maxDim * numbers.INITIAL_ORIGIN_SCALE;
        const maxRadius = surfaceDiameter + numbers.PADDING;
        const fgScale = maxRadius / initialSize;
        const unboundedCoords = unbounded ? {
            left: Math.round((width / 2) - (initialSize / 2)),
            top: Math.round((height / 2) - (initialSize / 2)),
        } : {left: 0, top: 0};

        let styles = {
            [strings.VAR_FG_SIZE]: `${initialSize}px`,
            [strings.VAR_FG_SCALE]: `${fgScale}`,
            [strings.VAR_LEFT]: `${unboundedCoords.left}px`,
            [strings.VAR_TOP]: `${unboundedCoords.top}px`
        };

        if (mode === RippleMode.activated || mode === RippleMode.deactivated) {
            let translateStart = '';
            let translateEnd = '';

            if (!unbounded) {
                const {startPoint, endPoint} = this.getFgTranslationCoordinates(initialSize);
                translateStart = `${startPoint.x}px, ${startPoint.y}px`;
                translateEnd = `${endPoint.x}px, ${endPoint.y}px`;
            }

            styles = Object.assign(styles, {
                [strings.VAR_FG_TRANSLATE_START]: translateStart,
                [strings.VAR_FG_TRANSLATE_END]: translateEnd
            });
        }

        let props: RippleComponentProps = {
            className: classNames,
            style: styles,
            onTouchStart: this.handleActivate,
            onMouseDown: this.handleActivate,
            onTouchEnd: this.handleDeactivate,
            onMouseUp: this.handleDeactivate,
            onFocus: this.handleFocus,
            onBlur: this.handleBlur,
        };

        if (!computeBoundingRect) {
            props = Object.assign(props, {ref: this.setInnerRef});
        }

        return this.props.render(props);
    }

    private getFgTranslationCoordinates = (initialSize: number) => {
        const {eventCoords, frame} = this.state;

        const startPoint = {
            x: eventCoords.x - (initialSize / 2),
            y: eventCoords.y - (initialSize / 2)
        };
        const endPoint = {
            x: (frame.width / 2) - (initialSize / 2),
            y: (frame.height / 2) - (initialSize / 2),
        };

        return {startPoint, endPoint};
    };

    private handleResize = () => {
        if (this._layoutFrame) {
            cancelAnimationFrame(this._layoutFrame);
        }

        this._layoutFrame = requestAnimationFrame(() => {
            const {width, height} = this.getBoundingClientRect();
            this.setState({frame: {width, height}});
            this._layoutFrame = 0;
        });
    };

    private handleActivate: EventHandler<React.MouseEvent<{}> | TouchEvent<{}>> = event => {
        if (this.props.disabled || this.state.mode === RippleMode.activated) {
            return;
        }

        this.setState({
            mode: RippleMode.activated,
            eventCoords: getNormalizedEventCoords(event, {
                x: window.pageXOffset,
                y: window.pageYOffset
            }, this.getBoundingClientRect())
        });

        document.documentElement.addEventListener('touchend', this.handleDeactivate, applyPassive());
        document.documentElement.addEventListener('mouseup', this.handleDeactivate, applyPassive());
        this._activationHasEnded = false;
    };

    private handleDeactivate = () => {
        if (this.state.mode === RippleMode.activated) {
            this._activationHasEnded = true;
            this.deactivate();
        }
    };

    private deactivate = () => {
        const {
            _activationAnimationHasEnded: activationAnimationHasEnded,
            _activationHasEnded: activationHasEnded
        } = this;

        if (activationAnimationHasEnded && activationHasEnded) {
            this._activationAnimationHasEnded = false;
            this.setState({mode: RippleMode.deactivated});

            document.documentElement.removeEventListener('touchend', this.handleDeactivate, applyPassive());
            document.documentElement.removeEventListener('mouseup', this.handleDeactivate, applyPassive());
        }
    };

    private handleFocus = () => this.setState({focused: true});
    private handleBlur = () => this.setState({focused: false});
    private setInnerRef = (ref: Element | null) => this._innerRef = ref;

    private getBoundingClientRect = (): ClientRect => {
        const {computeBoundingRect} = this.props;

        if (computeBoundingRect) {
            return computeBoundingRect();
        } else if (this._innerRef) {
            return this._innerRef.getBoundingClientRect();
        } else {
            return defaultBoundingRect;
        }
    };
}

export { Ripple as default, Ripple, RippleProps, RippleComponentProps, defaultBoundingRect };