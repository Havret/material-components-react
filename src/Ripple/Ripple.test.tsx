import { configure, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import toJson from 'enzyme-to-json';

import { Ripple, RippleProps } from './Ripple';

describe('Button', () => {
    configure({adapter: new Adapter()});

    const setUp = (props: Partial<RippleProps>, content: string = 'Ripple') => {
        const wrapper = mount(
            <Ripple
                {...props}
                render={({innerRef, ...rest}) =>
                    (<div {...rest} ref={innerRef}>{content}</div>)}
            />);
        return toJson(wrapper);
    };

    test('Bounded', () => {
        expect(setUp({})).toMatchSnapshot();
    });

    test('Unbounded', () => {
        expect(setUp({unbounded: true})).toMatchSnapshot();
    });

    test('Primary', () => {
        expect(setUp({primary: true})).toMatchSnapshot();
    });

    test('Secondary', () => {
        expect(setUp({secondary: true})).toMatchSnapshot();
    });
});