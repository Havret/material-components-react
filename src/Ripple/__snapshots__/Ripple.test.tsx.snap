// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Bounded 1`] = `
<Ripple
  disabled={false}
  primary={false}
  render={[Function]}
  secondary={false}
  surface={true}
  unbounded={false}
>
  <div
    className="mdc-ripple-upgraded mdc-ripple-surface"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "--mdc-ripple-fg-scale": "Infinity",
        "--mdc-ripple-fg-size": "0px",
        "--mdc-ripple-left": "0px",
        "--mdc-ripple-top": "0px",
      }
    }
  >
    Ripple
  </div>
</Ripple>
`;

exports[`Button Primary 1`] = `
<Ripple
  disabled={false}
  primary={true}
  render={[Function]}
  secondary={false}
  surface={true}
  unbounded={false}
>
  <div
    className="mdc-ripple-upgraded mdc-ripple-surface mdc-ripple-surface--primary"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "--mdc-ripple-fg-scale": "Infinity",
        "--mdc-ripple-fg-size": "0px",
        "--mdc-ripple-left": "0px",
        "--mdc-ripple-top": "0px",
      }
    }
  >
    Ripple
  </div>
</Ripple>
`;

exports[`Button Secondary 1`] = `
<Ripple
  disabled={false}
  primary={false}
  render={[Function]}
  secondary={true}
  surface={true}
  unbounded={false}
>
  <div
    className="mdc-ripple-upgraded mdc-ripple-surface mdc-ripple-surface--accent"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "--mdc-ripple-fg-scale": "Infinity",
        "--mdc-ripple-fg-size": "0px",
        "--mdc-ripple-left": "0px",
        "--mdc-ripple-top": "0px",
      }
    }
  >
    Ripple
  </div>
</Ripple>
`;

exports[`Button Unbounded 1`] = `
<Ripple
  disabled={false}
  primary={false}
  render={[Function]}
  secondary={false}
  surface={true}
  unbounded={true}
>
  <div
    className="mdc-ripple-upgraded mdc-ripple-upgraded--unbounded mdc-ripple-surface"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "--mdc-ripple-fg-scale": "NaN",
        "--mdc-ripple-fg-size": "0px",
        "--mdc-ripple-left": "0px",
        "--mdc-ripple-top": "0px",
      }
    }
  >
    Ripple
  </div>
</Ripple>
`;
